### PR TITLE
feature: publish in well-known and document private_key_jwt parameters in rest API

### LINKED ISSUE #2235: Support private_key_jwt from https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication
### What version of UAA are you running?

What output do you see from `curl <YOUR_UAA>/info -H'Accept: application/json'`?
> Develop / 76.x.x

### How are you deploying the UAA?

I am deploying the UAA

- locally only using gradlew
- using a bosh release I downloaded from bosh.io
- using cf-release
- using cf-deployment
- as part of a commercial Cloud Foundry distribution
- other (please explain)

### What did you do?

Authenticate OAuth clients without secrets. Public Flow does not solve all cases, e.g. client_credential or password grant .
Private_key_jwt and https://www.rfc-editor.org/rfc/rfc8705 would solve that issue.

Private_key_jwt is part of OIDC standard , e.g.
https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

And therefore this is Expected to be Part of UAA

### What did you expect to see? What goal are you trying to achieve with the UAA?

Private_key_jwt is part of OIDC standard , e.g.
https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

### What did you see instead?

secret based authentication only

### PR DESCRIPTION
**Pull request by strehle**
* Use of client_secret is optional
* Link to change JWT client

### CODE REVIEW

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/login/TokenEndpointDocs.java [hunk 1]
```diff
@@ -213,6 +220,8 @@ void getTokenUsingAuthCodeGrant() throws Exception {
                 parameterWithName("code").description(codeDescription).attributes(SnippetUtils.constraints.value("Required"), SnippetUtils.type.value(STRING)),
                 grantTypeParameter.description("the type of authentication being used to obtain the token, in this case `authorization_code`"),
                 clientSecretParameter.description("<small><mark>UAA 75.21.0</mark></small> Optional and can be omitted if client has configured allowpublic and [PKCE](https://tools.ietf.org/html/rfc7636) with `code_challange_method=S256` is used to create to `code`."),
```
Comment (adrianhoelzl-sap) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/login/TokenEndpointDocs.java]: It's not part of this PR, but shouldn't this be `code_challenge_method`?

**Review by adrianhoelzl-sap — APPROVED** [review]

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/login/TokenEndpointDocs.java (hunk 1 above)
Comment (strehle) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/login/TokenEndpointDocs.java]: Thanks for finding it, sure was wrong in past

**Comment by hsinn0** [comment]
When I saw the title of this PR starting with "documentation:", I assumed that the change is documentation only. However, it contains more than documentation change.

File: uaa/slateCustomizations/source/index.html.md.erb [hunk 2]
```diff
@@ -156,9 +156,12 @@ _Response Headers_
 
 # Token
 
-The `/oauth/token` endpoint requires client authentication to be accessed. Client Authentication can be passed as
-as part of the request authorization header, using basic authentication, or as part of the request parameters, using the `client_id` and `client_secret` parameter
-names.
+The `/oauth/token` endpoint requires client authentication to be accessed, except you allow the public usage. For more details about public usage see
+allowpublic flag in client details configuration. For confidential client usages the client authentication can be passed as part of the request
```
Comment (hsinn0) [inline uaa/slateCustomizations/source/index.html.md.erb]: I find having a `,` after `For confidential client usages` makes it more readable.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/JwtBearerGrantEndpointDocs.java [hunk 3]
```diff
@@ -76,7 +76,9 @@ void document_jwt_bearer_grant() throws Exception {
...
             parameterWithName("assertion").type(STRING).required().description("JWT token identifying representing the user to be authenticated"),
             parameterWithName("client_id").type(STRING).required().description("Required, client with "),
-            parameterWithName("client_secret").type(STRING).required().description("Required unless a basic authorization header is used"),
+            parameterWithName("client_secret").type(STRING).optional(null).description("The [secret passphrase configured](#change-secret) for the OAuth client. Optional if it is passed as part of the Basic Authorization header or if client_assertion is sent as part of private_key_jwt authentication."),
+            parameterWithName("client_assertion").type(STRING).optional(null).description("<small><mark>UAA 76.22.0</mark></small> Client authentication using method [private_key_jwt](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) optional as replacement of methods client_secret_basic or client_secret_post using secrets. The client needs to have a valid [JWT confiuration](#change-client-jwt) for trust to JWT in client_assertion."),
```
Comment (hsinn0) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/JwtBearerGrantEndpointDocs.java]: Is mentioning specific version "UAA 76.22.0" intentional?
Additionally, isn't there supposed to be a `.` between `core-1_0.html#ClientAuthentication)` and ` optional as replacement`, that is, aren't they supposed to be separate statements?

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/JwtBearerGrantEndpointDocs.java [hunk 4]
```diff
@@ -76,7 +76,9 @@ void document_jwt_bearer_grant() throws Exception {
...
+            parameterWithName("client_assertion_type").type(STRING).optional(null).description("<small><mark>UAA 76.22.0</mark></small> [RFC 7523](https://tools.ietf.org/html/rfc7523) describes the type. Must be set to `urn:ietf:params:oauth:client-assertion-type:jwt-bearer` if `client_assertion` parameter is present."),
```
Comment (hsinn0) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/JwtBearerGrantEndpointDocs.java]: Same question about "UAA 76.22.0" mark.

**Comment by strehle** [comment]
> When I saw the title of this PR starting with "documentation:", I assumed that the change is documentation only. However, it contains more than documentation change.

sorry, yes I started with documentation only but then found missing well-known adoption which belongs to the story, so should I devide this PR or rename PR ?

**Comment by strehle** [comment]
Thanks for your review

**Comment by hsinn0** [comment]
> > When I saw the title of this PR starting with "documentation:", I assumed that the change is documentation only. However, it contains more than documentation change.
> 
> sorry, yes I started with documentation only but then found missing well-known adoption which belongs to the story, so should I devide this PR or rename PR ?

Looks like u renamed it, so it is good now.

**Review by hsinn0 — APPROVED** [review]
